gobblin-core-base/src/main/java/gobblin/writer/AsyncWriterManager.java

```diff
@@ -0,0 +1,519 @@
+/*
+ * Copyright (C) 2014-2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.writer;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+import com.codahale.metrics.Meter;
+import com.codahale.metrics.Timer;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import com.google.common.collect.Lists;
+import com.google.common.io.Closer;
+import com.typesafe.config.Config;
+import com.typesafe.config.ConfigFactory;
+
+import javax.annotation.Nonnull;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+
+import gobblin.configuration.State;
+import gobblin.instrumented.Instrumentable;
+import gobblin.instrumented.Instrumented;
+import gobblin.metrics.GobblinMetrics;
+import gobblin.metrics.MetricContext;
+import gobblin.metrics.MetricNames;
+import gobblin.metrics.Tag;
+import gobblin.util.ConfigUtils;
+import gobblin.util.FinalState;
+
+
+/**
+ * A Data Writer to use as a base for writing async writers.
+ * This Data Writer wraps writers that implement the {@link AsyncDataWriter} interface
+ * and provides the following features:
+ * 1. Calculate metrics for: number of records in, attempted, successfully written, failed, latency.
+ * 2. Wait for a specified amount of time on commit for all pending writes to complete.
+ * 3. Do not proceed if a certain failure threshold is exceeded.
+ * 4. Support a fixed number of retries on failure of individual records (TODO: retry strategies)
+ * 5. Support a max number of outstanding / unacknowledged writes
+ * 6. TODO: Support ordered / unordered write semantics
+ *
+ *
+ */
+@Slf4j
```

Comment:
Should I then remove the @slf4j annotation? 
Is the recommendation to hand create a logger if one isn't passed down?